File: backend/app/clients/congress.py
import logging
from typing import Optional

from .base import BaseAPIClient
from ..config import get_settings
from ..models.schemas import SourceItem

logger = logging.getLogger(__name__)


class CongressClient(BaseAPIClient):
    def __init__(self):
        settings = get_settings()
        super().__init__(base_url=settings.congress_base_url)
        self.api_key = settings.gov_api_key

    def _get_headers(self) -> dict:
        return {
            "X-Api-Key": self.api_key,
            "Accept": "application/json",
        }

    def _normalize_bill(self, bill: dict) -> SourceItem:
        bill_type = bill.get("type", "").lower()
        bill_number = bill.get("number", "")
        congress = bill.get("congress", "")

        url = f"https://www.congress.gov/bill/{congress}th-congress/{bill_type}/{bill_number}"

        update_date = bill.get("updateDate") or bill.get("latestAction", {}).get("actionDate")

        summary = None
        if bill.get("latestAction"):
            summary = bill["latestAction"].get("text", "")

        return SourceItem(
            source_type="congress_bill",
            id=f"{congress}-{bill_type}-{bill_number}",
            title=bill.get("title", "Untitled Bill"),
            agency=None,
            date=update_date,
            url=url,
            excerpt=summary,
            content_type="bill",
            raw=bill,
        )

    def _normalize_vote(self, vote: dict, chamber: str = "house") -> SourceItem:
        vote_number = vote.get("rollNumber") or vote.get("rollCallNumber", "")
        congress = vote.get("congress", "")
        session = vote.get("session", "")

        url = f"https://www.congress.gov/roll-call-vote/{congress}th-congress-{session}/{chamber}/{vote_number}"

        question = vote.get("question", "Roll Call Vote")
        result = vote.get("result", "")
        title = f"{question}" + (f" - {result}" if result else "")

        vote_date = vote.get("date") or vote.get("updateDate")

        return SourceItem(
            source_type="congress_vote",
            id=f"{congress}-{session}-{chamber}-{vote_number}",
            title=title,
            agency=chamber.capitalize(),
            date=vote_date,
            url=url,
            excerpt=vote.get("description"),
            content_type="vote",
            raw=vote,
        )

    async def search_bills(
        self,
        query: str,
        congress: Optional[int] = None,
        bill_type: Optional[str] = None,
        limit: int = 10,
        offset: int = 0,
    ) -> tuple[list[dict], list[SourceItem]]:
        endpoint = "/bill"
        if congress:
            endpoint = f"/bill/{congress}"

        params = {
            "format": "json",
            "limit": min(limit, 250),
            "offset": offset,
        }

        url = f"{self.base_url}{endpoint}"
        logger.info(f"Fetching Congress bills: {url}")

        data = await self._request_with_retry(
            method="GET",
            url=url,
            headers=self._get_headers(),
            params=params,
        )

        bills = data.get("bills", [])

        if query:
            query_lower = query.lower()
            bills = [
                b for b in bills
                if query_lower in (b.get("title", "") or "").lower()
                or query_lower in (b.get("number", "") or "").lower()
            ]

        sources = [self._normalize_bill(bill) for bill in bills[:limit]]
        return bills[:limit], sources

    async def get_bill(
        self,
        congress: int,
        bill_type: str,
        bill_number: int,
    ) -> tuple[Optional[dict], Optional[SourceItem]]:
        url = f"{self.base_url}/bill/{congress}/{bill_type.lower()}/{bill_number}"

        params = {"format": "json"}

        logger.info(f"Fetching Congress bill: {url}")

        data = await self._request_with_retry(
            method="GET",
            url=url,
            headers=self._get_headers(),
            params=params,
        )

        bill = data.get("bill", {})
        if bill:
            return bill, self._normalize_bill(bill)
        return None, None

    async def search_votes(
        self,
        chamber: str = "house",
        congress: Optional[int] = None,
        limit: int = 10,
    ) -> tuple[list[dict], list[SourceItem]]:
        if not congress:
            congress = 118

        endpoint = f"/{chamber}/rollCall/{congress}"
        url = f"{self.base_url}{endpoint}"

        params = {
            "format": "json",
            "limit": min(limit, 250),
        }

        logger.info(f"Fetching Congress votes: {url}")

        data = await self._request_with_retry(
            method="GET",
            url=url,
            headers=self._get_headers(),
            params=params,
        )

        votes = data.get("roll_calls", data.get("rollCalls", []))
        if not votes and "vote" in data:
            votes = [data["vote"]]

        sources = [self._normalize_vote(vote, chamber) for vote in votes[:limit]]
        return votes[:limit], sources

File: backend-ts/src/clients/congress.ts
import { BaseAPIClient } from "./base.js";
import { getSettings } from "../config.js";
import type { SourceItem } from "../models/schemas.js";

export class CongressClient extends BaseAPIClient {
    private apiKey: string;

    constructor() {
        const settings = getSettings();
        super(settings.congressBaseUrl);
        this.apiKey = settings.govApiKey;
    }

    private getHeaders(): Record<string, string> {
        return {
            "X-Api-Key": this.apiKey,
            Accept: "application/json",
        };
    }

    private normalizeBill(bill: Record<string, unknown>): SourceItem {
        const billType = ((bill.type as string) || "").toLowerCase();
        const billNumber = bill.number as string || "";
        const congress = bill.congress as number | string || "";

        const url = `https://www.congress.gov/bill/${congress}th-congress/${billType}/${billNumber}`;

        const latestAction = bill.latestAction as Record<string, unknown> | undefined;
        const updateDate =
            (bill.updateDate as string) || latestAction?.actionDate as string || null;

        const summary = latestAction?.text as string || null;

        return {
            source_type: "congress_bill",
            id: `${congress}-${billType}-${billNumber}`,
            title: (bill.title as string) || "Untitled Bill",
            agency: null,
            date: updateDate,
            url,
            excerpt: summary,
            content_type: "bill",
            raw: bill,
        };
    }

    private normalizeVote(vote: Record<string, unknown>, chamber: string = "house"): SourceItem {
        const voteNumber =
            (vote.rollNumber as string) || (vote.rollCallNumber as string) || "";
        const congress = vote.congress as string || "";
        const session = vote.session as string || "";

        const url = `https://www.congress.gov/roll-call-vote/${congress}th-congress-${session}/${chamber}/${voteNumber}`;

        const question = (vote.question as string) || "Roll Call Vote";
        const voteResult = vote.result as string || "";
        const title = voteResult ? `${question} - ${voteResult}` : question;

        const voteDate = (vote.date as string) || (vote.updateDate as string) || null;

        return {
            source_type: "congress_vote",
            id: `${congress}-${session}-${chamber}-${voteNumber}`,
            title,
            agency: chamber.charAt(0).toUpperCase() + chamber.slice(1),
            date: voteDate,
            url,
            excerpt: vote.description as string || null,
            content_type: "vote",
            raw: vote,
        };
    }

    async searchBills(options: {
        query: string;
        congress?: number;
        billType?: string;
        limit?: number;
        offset?: number;
    }): Promise<{ bills: Record<string, unknown>[]; sources: SourceItem[] }> {
        const { query, congress, limit = 10, offset = 0 } = options;

        let endpoint = "/bill";
        if (congress) {
            endpoint = `/bill/${congress}`;
        }

        const params: Record<string, string | number> = {
            format: "json",
            limit: Math.min(limit, 250),
            offset,
        };

        const url = `${this.baseUrl}${endpoint}`;
        console.log(`Fetching Congress bills: ${url}`);

        const data = await this.requestWithRetry<{ bills: Record<string, unknown>[] }>({
            url,
            headers: this.getHeaders(),
            params,
        });

        let bills = data.bills || [];

        // Client-side filter by query
        if (query) {
            const queryLower = query.toLowerCase();
            bills = bills.filter(
                (b) =>
                    ((b.title as string) || "").toLowerCase().includes(queryLower) ||
                    ((b.number as string) || "").toLowerCase().includes(queryLower)
            );
        }

        bills = bills.slice(0, limit);
        const sources = bills.map((bill) => this.normalizeBill(bill));

        return { bills, sources };
    }

    async getBill(options: {
        congress: number;
        billType: string;
        billNumber: number;
    }): Promise<{ bill: Record<string, unknown> | null; source: SourceItem | null }> {
        const { congress, billType, billNumber } = options;

        const url = `${this.baseUrl}/bill/${congress}/${billType.toLowerCase()}/${billNumber}`;
        const params = { format: "json" };

        console.log(`Fetching Congress bill: ${url}`);

        const data = await this.requestWithRetry<{ bill: Record<string, unknown> }>({
            url,
            headers: this.getHeaders(),
            params,
        });

        const bill = data.bill;
        if (bill) {
            return { bill, source: this.normalizeBill(bill) };
        }
        return { bill: null, source: null };
    }

    async searchVotes(options: {
        chamber?: string;
        congress?: number;
        limit?: number;
    }): Promise<{ votes: Record<string, unknown>[]; sources: SourceItem[] }> {
        const { chamber = "house", congress = 118, limit = 10 } = options;

        const endpoint = `/${chamber}/rollCall/${congress}`;
        const url = `${this.baseUrl}${endpoint}`;

        const params: Record<string, string | number> = {
            format: "json",
            limit: Math.min(limit, 250),
        };

        console.log(`Fetching Congress votes: ${url}`);

        const data = await this.requestWithRetry<Record<string, unknown>>({
            url,
            headers: this.getHeaders(),
            params,
        });

        let votes =
            (data.roll_calls as Record<string, unknown>[]) ||
            (data.rollCalls as Record<string, unknown>[]) ||
            [];

        if (!votes.length && data.vote) {
            votes = [data.vote as Record<string, unknown>];
        }

        votes = votes.slice(0, limit);
        const sources = votes.map((vote) => this.normalizeVote(vote, chamber));

        return { votes, sources };
    }
}
